t/bc/bc.t

```diff
@@ -0,0 +1,207 @@
+        [ 'v = -(4+5)+8', '-1',      'PEDMAS' ],
+    ];
+    return $table;
+}
+
+sub special_expr_table {
+    my $sqrt2 = '1.4142135623731';
+    my $table = [
+        'special expressions',
+        [ 'sqrt(2)',          $sqrt2,     'sqrt function' ],
+        [ 'length(sqrt(2))',  '14',       'significant digits' ],
+        [ 'scale(sqrt(2))',   '13',       'precision digits' ],
+    ];
+    return $table;
+}
+
+sub statement_table {
+    ## no critic [ValuesAndExpressions::RequireInterpolationOfMetachars]
+    my $table = [
+        'statements',
+        [    q("a string"),
+            'a string',
+            'string literal',
+        ],
+
+        [
+            # need to print the newline here so that the test harness
+            # recognizes the end of the test output
+            'print "1+2", " is ", 1+2, "\n"',
+            '1+2 is 3',
+            'print statement',
+        ],
+
+        [   '{ print "a"; print "b"; print "\n" }',
+            'ab',
+            'compound statement',
+        ],
+
+        [   'v=5; while (v--) { print v }; print "\n"',
+            "5\n43210",
+            'while statement',
+        ],
+
+        ## no critic [ValuesAndExpressions::ProhibitInterpolationOfLiterals]
+        [   'for (v=0;v<5;v++) { print v }; print "\n"',
+            "01234",
+            'for statement',
+        ],
+
+        [   'for ( v=0; v<5; v++) { print v; if (v>2) break }; print "\n"',
+            "01230",
+            'for with break statement',
+        ],
+
+        # The following aren't being tested because they are extensions
+        # that aren't support in the PerlPowerTools implementation of bc
+        # - read()
+        # - if ... else
+        # - continue
+        # - halt
+        # - warranty
+        # - define (because I can't even get it to work interactively)
+
+        # Not testing this, for obvious reasons
+        # - quit
+    ];
+    return $table;
+}
+
+sub run_table {
+    my ($table) = @_;
+
+    my $label = shift @$table;
+
+    subtest $label => sub {
+        my ( $fh, $input ) = tempfile();
+        foreach my $tuple (@$table) {
+            my ( $expr, $expected, $desc ) = @$tuple;
+            print {$fh} "$expr\n" or die;
+        }
+        close $fh or die;
+
+        ## no critic [InputOutput::ProhibitBacktickOperators]
+        my $output = `perl $Script $input`;
+        my @got    = split /\n/xms, $output;
+
+        # get expected results
+        my @expected;
+        my @message;
+        foreach my $t_ar (@$table) {
+            my ( $expr, $exp, $desc ) = @{$t_ar};
+
+            # some operations generate multiple lines of output
+            my @lines = split /\n/xms, $exp;
+            foreach my $ln (@lines) {
+                push @expected, $ln;
+                push @message,  $desc . ' : ' . $expr;
+            }
+        }
+
+        # is @got, @expected, 'count of results';
+
+        foreach my $got (@got) {
```

Comment:
No, I can't test them as as I loop through them without re-executing bc ever single time.  That was my previous solution and after benchmarking it against this solution the running time was 4 seconds vs 0.5 seconds.  And I was doing a lot fewer tests.  This strategy takes advantage of the fact that bc can take a file as input, so I write out the expression to a temp file then run bc once.  The catch is that bc sometimes outputs more than one line per result.  Hence the newlines in the expected column of the table.  

I wouldn't normally do it this way, but for bc it's a faster solution than executing the script over and over again and it avoids the use of Test::Cmd.  You'll note that bc is re-execute for each table, just not within each table.  That gives plenty of leeway to group tests so that a failure in one group doesn't cascade to the others.

Reply:
Changes pushed.  I didn't change the approach, but I added some comments to explain why it's being done that way.